File: src/ros2_bridge/mappers/shape_msgs/mesh_triangle.rs
//! Typed mapper for `shape_msgs/msg/MeshTriangle`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn mesh_triangle_to_bus(msg: ros_env::shape_msgs::msg::MeshTriangle) -> crate::shape_msgs::msg::v1::MeshTriangle {
    crate::shape_msgs::msg::v1::MeshTriangle {
        vertex_indices: crate::ros2_bridge::mappers::convert::u32_seq(msg.vertex_indices),
    }
}

pub(crate) fn mesh_triangle_to_ros(bus: crate::shape_msgs::msg::v1::MeshTriangle) -> ros_env::shape_msgs::msg::MeshTriangle {
    ros_env::shape_msgs::msg::MeshTriangle {
        vertex_indices: crate::ros2_bridge::mappers::convert::FromU32Seq::from_u32_seq(bus.vertex_indices),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ShapeMsgsMeshTriangleMapper;

impl TypedTopicMapper for ShapeMsgsMeshTriangleMapper {
    type Ros = ros_env::shape_msgs::msg::MeshTriangle;
    type Bus = crate::shape_msgs::msg::v1::MeshTriangle;

    fn type_name(&self) -> &'static str {
        "shape_msgs/msg/MeshTriangle"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(mesh_triangle_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(mesh_triangle_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/shape_msgs/mesh_triangle.py
"""Generated mapper for `shape_msgs/msg/MeshTriangle`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.shape_msgs.msg.v1 import MeshTriangle as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def mesh_triangle_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.vertex_indices.extend(list(msg.vertex_indices))
    return bus


def mesh_triangle_to_ros(bus):
    from shape_msgs.msg import MeshTriangle as RosMsg

    out = RosMsg()
    out.vertex_indices = list(bus.vertex_indices)
    return out


class ShapeMsgsMeshTriangleMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from shape_msgs.msg import MeshTriangle as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return mesh_triangle_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return mesh_triangle_to_ros(bus)
